commit 0234a307540ee2373aee5260695f8d0c55849556
Author: Susan Franklin <sfranklin>
Date:   Tue Apr 28 23:51:38 2009 +0000

    Bug 274096 - [ui] Update Dialog - Next button enabled but does not function after failed resolution

1	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
index 5389894..c6c4223 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
@@ -146,6 +146,7 @@ public abstract class ResolutionStatusPage extends ProvisioningWizardPage {
 				detail = resolvedOperation.getResolutionResult().getSummaryReport();
 				detailsGroup.enablePropertyLink(false);
 				detailsGroup.getDetailsArea().setText(detail);
+				return;
 			}
 
 			// The overall status is not an error, so we may as well just show info about this iu rather than everything.
commit 1a5eeac946d9c3d64c7d8cd6e50d851dca3e2d11
Author: Susan Franklin <sfranklin>
Date:   Wed Apr 29 21:55:57 2009 +0000

    Bug 274096 - [ui] Update Dialog - Next button enabled but does not function after failed resolution

3	5	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUsPage.java
9	3	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/IUDetailsGroup.java
10	2	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
0	2	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionResultsWizardPage.java
5	5	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
1	3	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUsPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUsPage.java
index c306cb7..76286b9 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUsPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/AvailableIUsPage.java
@@ -59,7 +59,6 @@ public class AvailableIUsPage extends ProvisioningWizardPage implements ISelecta
 	Composite availableIUButtonBar;
 	Link installLink;
 	Button useCategoriesCheckbox, hideInstalledCheckbox, showLatestVersionsCheckbox, resolveAllCheckbox;
-	Text detailsArea;
 	SashForm sashForm;
 	IUColumnConfig nameColumn, versionColumn;
 	StructuredViewerProvisioningListener profileListener;
@@ -138,7 +137,6 @@ public class AvailableIUsPage extends ProvisioningWizardPage implements ISelecta
 
 		// Details area
 		iuDetailsGroup = new IUDetailsGroup(sashForm, availableIUGroup.getStructuredViewer(), SWT.DEFAULT, true);
-		detailsArea = iuDetailsGroup.getDetailsArea();
 
 		sashForm.setWeights(getSashWeights());
 
@@ -452,7 +450,7 @@ public class AvailableIUsPage extends ProvisioningWizardPage implements ISelecta
 		if (elements.length == 1 && elements[0] instanceof EmptyElementExplanation) {
 			String description = ((EmptyElementExplanation) elements[0]).getDescription();
 			if (description != null) {
-				detailsArea.setText(description);
+				iuDetailsGroup.setDetailText(description);
 				return;
 			}
 		}
@@ -474,10 +472,10 @@ public class AvailableIUsPage extends ProvisioningWizardPage implements ISelecta
 				result.append(selected[0].getVersion().toString());
 			}
 
-			detailsArea.setText(result.toString());
+			iuDetailsGroup.setDetailText(result.toString());
 			return;
 		}
-		detailsArea.setText(""); //$NON-NLS-1$
+		iuDetailsGroup.setDetailText(""); //$NON-NLS-1$
 	}
 
 	public IInstallableUnit[] getSelectedIUs() {
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/IUDetailsGroup.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/IUDetailsGroup.java
index 1f3e3a6..733c36c 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/IUDetailsGroup.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/IUDetailsGroup.java
@@ -39,6 +39,7 @@ public class IUDetailsGroup {
 	private ISelectionProvider selectionProvider;
 	private int widthHint;
 	private boolean scrollable;
+	private String lastText;
 
 	/**
 	 * 
@@ -91,10 +92,15 @@ public class IUDetailsGroup {
 	}
 
 	/**
-	 * @return The details area for this IUDEtailsGroup
+	 * Set the detail text
 	 */
-	public Text getDetailsArea() {
-		return this.detailsArea;
+	public void setDetailText(String text) {
+		// If the string is the same but the user has scrolled, the text
+		// widget will reset the selection.  This makes it look like the text
+		// has changed when it hasn't.  For this reason, we check equality first.
+		if (lastText == null || !lastText.equals(text))
+			detailsArea.setText(text);
+		lastText = text;
 	}
 
 	/**
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
index 72a1a15..fbd292c 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ProvisioningOperationWizard.java
@@ -113,6 +113,7 @@ public abstract class ProvisioningOperationWizard extends Wizard {
 		if (page == mainPage || page == errorPage) {
 			ISelectableIUsPage currentPage = (ISelectableIUsPage) page;
 			// Do we need to resolve?
+			boolean weResolved = false;
 			if (resolutionOperation == null || (resolutionOperation != null && shouldRecomputePlan(currentPage))) {
 				resolutionOperation = null;
 				provisioningContext = getProvisioningContext();
@@ -120,21 +121,28 @@ public abstract class ProvisioningOperationWizard extends Wizard {
 				root = makeResolutionElementRoot(planSelections);
 				recomputePlan(getContainer());
 				planChanged();
+				weResolved = true;
 			} else {
 				planSelections = currentPage.getCheckedIUElements();
 				root = makeResolutionElementRoot(planSelections);
 			}
-			return selectNextPage(page, getCurrentStatus());
+			return selectNextPage(page, getCurrentStatus(), weResolved);
 		}
 		return super.getNextPage(page);
 	}
 
-	protected IWizardPage selectNextPage(IWizardPage currentPage, IStatus status) {
+	protected IWizardPage selectNextPage(IWizardPage currentPage, IStatus status, boolean hasResolved) {
+		// We have already established before calling this method that the
+		// current page is either the main page or the error page.  
 		if (status.getSeverity() == IStatus.CANCEL)
 			return currentPage;
 		else if (status.getSeverity() == IStatus.ERROR) {
 			if (errorPage == null)
 				errorPage = getErrorReportingPage();
+			if (currentPage == errorPage) {
+				updateErrorPageStatus(errorPage);
+				return null;
+			}
 			showingErrorPage();
 			return errorPage;
 		} else {
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionResultsWizardPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionResultsWizardPage.java
index 41b4e4b..f788896 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionResultsWizardPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionResultsWizardPage.java
@@ -52,7 +52,6 @@ public abstract class ResolutionResultsWizardPage extends ResolutionStatusPage {
 	PlannerResolutionOperation resolvedOperation;
 	protected Policy policy;
 	TreeViewer treeViewer;
-	Text detailsArea;
 	ProvElementContentProvider contentProvider;
 	IUDetailsLabelProvider labelProvider;
 	protected Display display;
@@ -131,7 +130,6 @@ public abstract class ResolutionResultsWizardPage extends ResolutionStatusPage {
 
 		// The text area shows a description of the selected IU, or error detail if applicable.
 		iuDetailsGroup = new IUDetailsGroup(sashForm, treeViewer, convertWidthInCharsToPixels(ILayoutConstants.DEFAULT_TABLE_WIDTH), true);
-		detailsArea = iuDetailsGroup.getDetailsArea();
 
 		updateStatus(input, resolvedOperation);
 		setControl(sashForm);
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
index c6c4223..cf59339 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/ResolutionStatusPage.java
@@ -129,7 +129,7 @@ public abstract class ResolutionStatusPage extends ProvisioningWizardPage {
 				detail = ""; //$NON-NLS-1$
 				detailsGroup.enablePropertyLink(false);
 			}
-			detailsGroup.getDetailsArea().setText(detail);
+			detailsGroup.setDetailText(detail);
 			return;
 		}
 
@@ -138,20 +138,20 @@ public abstract class ResolutionStatusPage extends ProvisioningWizardPage {
 			detail = resolvedOperation.getResolutionResult().getDetailedReport(new IInstallableUnit[] {selectedIU});
 			if (detail != null) {
 				detailsGroup.enablePropertyLink(false);
-				detailsGroup.getDetailsArea().setText(detail);
+				detailsGroup.setDetailText(detail);
 				return;
 			}
 			// No specific error about this IU.  Show the overall error if it is in error.
 			if (resolvedOperation.getResolutionResult().getSummaryStatus().getSeverity() == IStatus.ERROR) {
 				detail = resolvedOperation.getResolutionResult().getSummaryReport();
 				detailsGroup.enablePropertyLink(false);
-				detailsGroup.getDetailsArea().setText(detail);
+				detailsGroup.setDetailText(detail);
 				return;
 			}
 
 			// The overall status is not an error, so we may as well just show info about this iu rather than everything.
 			detailsGroup.enablePropertyLink(true);
-			detailsGroup.getDetailsArea().setText(getIUDescription(selectedIU));
+			detailsGroup.setDetailText(getIUDescription(selectedIU));
 			return;
 		}
 
@@ -160,7 +160,7 @@ public abstract class ResolutionStatusPage extends ProvisioningWizardPage {
 		detailsGroup.enablePropertyLink(false);
 		if (detail == null)
 			detail = ""; //$NON-NLS-1$
-		detailsGroup.getDetailsArea().setText(detail);
+		detailsGroup.setDetailText(detail);
 	}
 
 	protected abstract String getDialogSettingsName();
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
index 499a629..fd6a452 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/dialogs/SelectableIUsPage.java
@@ -48,7 +48,6 @@ public class SelectableIUsPage extends ResolutionStatusPage implements IResoluti
 	Object[] initialSelections;
 	PlannerResolutionOperation resolvedOperation;
 	CheckboxTableViewer tableViewer;
-	Text detailsArea;
 	IUDetailsGroup iuDetailsGroup;
 	ProvElementContentProvider contentProvider;
 	IUDetailsLabelProvider labelProvider;
@@ -137,7 +136,6 @@ public class SelectableIUsPage extends ResolutionStatusPage implements IResoluti
 		// The text area shows a description of the selected IU, or error detail if applicable.
 		iuDetailsGroup = new IUDetailsGroup(sashForm, tableViewer, convertWidthInCharsToPixels(ILayoutConstants.DEFAULT_TABLE_WIDTH), true);
 
-		detailsArea = iuDetailsGroup.getDetailsArea();
 		updateStatus(root, resolvedOperation);
 		setControl(sashForm);
 		sashForm.setWeights(getSashWeights());
@@ -280,7 +278,7 @@ public class SelectableIUsPage extends ResolutionStatusPage implements IResoluti
 			setPageComplete(false);
 			if (!isCreated())
 				return;
-			getDetailsGroup().getDetailsArea().setText(specialStatus.getMessage());
+			getDetailsGroup().setDetailText(specialStatus.getMessage());
 			setMessage(getMessageText(specialStatus), IMessageProvider.ERROR);
 
 		}
